### PR TITLE
[PS-1402] New RealPage Api

### PR DESCRIPTION
**Pull request by JaredCowan**
Silence is golden

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -8,3 +8,6 @@ composer.lock
 
 ### NetBeans
 nbproject
+
+### Misc
+.DS_Store
```
Comment (adam-winfree) [inline .gitignore:13]: I just put this in my global `.gitignore` but I can see the value of having it here.

File: src/Intacct/ClientConfig.php [hunk 2]
```diff
@@ -254,6 +254,42 @@ public function setUserPassword(string $userPassword)
...
+     */
+    public function getHeaders()
+    {
+        return $this->headers;
+    }
+
+    /**
+     * @param array $headers
+     */
+    public function setHeaders(array $headers)
```
Comment (VirgilAnderson) [inline src/Intacct/ClientConfig.php:270]: return type?

File: src/Intacct/ClientConfig.php [hunk 3]
```diff
@@ -254,6 +254,42 @@ public function setUserPassword(string $userPassword)
...
+    {
+        $this->headers = $headers;
+    }
+
+    /**
+     * @param string $key
+     * @param string $value
+     */
+    public function addHeader(string $key, string $value)
```
Comment (VirgilAnderson) [inline src/Intacct/ClientConfig.php:279]: return type

File: src/Intacct/ClientConfig.php [hunk 4]
```diff
@@ -254,6 +254,42 @@ public function setUserPassword(string $userPassword)
...
+    {
+        $this->headers[$key] = $value;
+    }
+
+    public function needsSenderId() : bool
+    {
+        $headerKeys = array_map(function ($key) {
+            return strtolower($key ?? '');
+        }, array_keys($this->headers));
+
```
Comment (adam-winfree) [inline src/Intacct/ClientConfig.php:289]: I've never thought that null could be a key. It looks like it would be cast to an empty string like you have here. https://www.php.net/manual/en/language.types.array.php#language.types.array.key-casts

If that's the case, you could shorten this to `$headerKeys = array_map('strtolower', array_keys($this->headers));`

File: src/Intacct/Functions/AccountsPayable/AbstractVendor.php [hunk 5]
```diff
@@ -1449,7 +1513,7 @@ public function getRestrictionType()
      *
      * @param string $restrictionType
      */
-    public function setRestrictionType($restrictionType)
+    public function setRestrictionType(string $restrictionType)
```
Comment (VirgilAnderson) [inline src/Intacct/Functions/AccountsPayable/AbstractVendor.php:1516]: return types here too?

File: src/Intacct/Functions/AccountsPayable/AbstractVendor.php [hunk 6]
```diff
@@ -1502,6 +1581,271 @@ public function setRestrictedDepartments($restrictedDepartments)
...
+
+    /**
+     * Set vendor designations (DEI)
+     *
+     * @param array $vendorDesignations
+     */
+    public function setVendorDesignations(array $vendorDesignations) : void
+    {
+        foreach ($vendorDesignations as $vendorDesignation) {
+            if (!isset($vendorDesignation['name'])) {
```
Comment (VirgilAnderson) [inline src/Intacct/Functions/AccountsPayable/AbstractVendor.php:1616]:
```suggestion
            if (! isset($vendorDesignation['name'])) {
```

File: src/Intacct/Functions/AccountsPayable/AbstractVendor.php [hunk 7]
```diff
@@ -1502,6 +1581,271 @@ public function setRestrictedDepartments($restrictedDepartments)
...
+    }
+
+    /**
+     * Add vendor designation (DEI)
+     *
+     * @param array $vendorDesignation
+     */
+    public function addVendorDesignation(array $vendorDesignation) : void
+    {
+        if (!isset($vendorDesignation['name'])) {
```
Comment (VirgilAnderson) [inline src/Intacct/Functions/AccountsPayable/AbstractVendor.php:1601]:
```suggestion
        if (! isset($vendorDesignation['name'])) {
```

File: src/Intacct/Functions/AccountsPayable/AbstractVendor.php [hunk 8]
```diff
@@ -1502,6 +1581,271 @@ public function setRestrictedDepartments($restrictedDepartments)
...
+            }
+        }
+
+        $this->vendorDesignations = $vendorDesignations;
+    }
+
+    /**
+     * @return mixed
+     */
+    public function getInsuranceRequired() {
```
Comment (VirgilAnderson) [inline src/Intacct/Functions/AccountsPayable/AbstractVendor.php:1627]: return types

File: src/Intacct/Functions/AccountsPayable/VendorCreate.php [hunk 9]
```diff
@@ -147,14 +148,44 @@ public function writeXml(XMLWriter &$xml)
         // TODO: Location assigned account numbers
 
         $xml->writeElement('OBJECTRESTRICTION', $this->getRestrictionType());
-        if (count($this->getRestrictedLocations()) > 0) {
+        if (!empty($this->getRestrictedLocations())) {
```
Comment (VirgilAnderson) [inline src/Intacct/Functions/AccountsPayable/VendorCreate.php:151]:
```suggestion
        if (! empty($this->getRestrictedLocations())) {
```

File: src/Intacct/Functions/AccountsPayable/VendorCreate.php [hunk 10]
```diff
@@ -147,14 +148,44 @@ public function writeXml(XMLWriter &$xml)
...
+
         if (count($this->getRestrictedDepartments()) > 0) {
             $xml->writeElement('RESTRICTEDDEPARTMENTS', $this->getRestrictedDepartments());
         }
 
+        if (!empty($this->getVendorDesignations())) {
+            $this->writeXmlVendorDesignations($xml);
+        }
+
+        if (!empty($this->getCustomFieldsV2())) {
```
Comment (VirgilAnderson) [inline src/Intacct/Functions/AccountsPayable/VendorCreate.php:163]:
```suggestion
        if (! empty($this->getCustomFieldsV2())) {
```

File: src/Intacct/Functions/AccountsPayable/VendorUpdate.php [hunk 11]
```diff
@@ -138,13 +139,42 @@ public function writeXml(XMLWriter &$xml)
         // TODO: Location assigned account numbers
 
         $xml->writeElement('OBJECTRESTRICTION', $this->getRestrictionType());
-        if (count($this->getRestrictedLocations()) > 0) {
+        if (!empty($this->getRestrictedLocations())) {
```
Comment (VirgilAnderson) [inline src/Intacct/Functions/AccountsPayable/VendorUpdate.php:142]:
```suggestion
        if (! empty($this->getRestrictedLocations())) {
```

File: src/Intacct/Functions/AccountsPayable/VendorUpdate.php [hunk 12]
```diff
@@ -138,13 +139,42 @@ public function writeXml(XMLWriter &$xml)
...
+
         if (count($this->getRestrictedDepartments()) > 0) {
             $xml->writeElement('RESTRICTEDDEPARTMENTS', $this->getRestrictedDepartments());
         }
 
+        if (!empty($this->getVendorDesignations())) {
+            $this->writeXmlVendorDesignations($xml);
+        }
+
+        if (!empty($this->getCustomFieldsV2())) {
```
Comment (VirgilAnderson) [inline src/Intacct/Functions/AccountsPayable/VendorUpdate.php:154]:
```suggestion
        if (! empty($this->getCustomFieldsV2())) {
```

File: src/Intacct/Functions/Traits/CustomFieldsV2Trait.php [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+    
+    /** @var array */
+    protected $customFieldsV2 = [];
+
+    /**
+     * Get custom fields
+     *
+     * @return array
+     */
+    public function getCustomFieldsV2(): array
```
Comment (VirgilAnderson) [inline src/Intacct/Functions/Traits/CustomFieldsV2Trait.php:23]:
```suggestion
    public function getCustomFieldsV2() : array
```

File: src/Intacct/Functions/Traits/CustomFieldsV2Trait.php [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+    {
+        return $this->customFieldsV2;
+    }
+
+    /**
+     * Set custom fields
+     *
+     * @param array $customFields
+     */
+    public function setCustomFieldsV2(array $customFields)
```
Comment (VirgilAnderson) [inline src/Intacct/Functions/Traits/CustomFieldsV2Trait.php:33]: return types

**Review by VirgilAnderson — COMMENTED** [review]
Couple little things

File: src/Intacct/ClientConfig.php [hunk 15]
```diff
@@ -254,6 +254,42 @@ public function setUserPassword(string $userPassword)
...
+        return ! empty($this->headers) && array_key_exists('apikey', $headerKeys);
```
Comment (adam-winfree) [inline src/Intacct/ClientConfig.php:290]: It seems like this could just be `in_array('apikey', $headerKeys)` or `array_key_exists('apikey', $headerKeys)`. It's not empty if it has the `apikey` key.

File: .gitignore (hunk 1 above)
Comment (JaredCowan) [reply .gitignore:13]: Not a big fan of globals because you need to assume everyone has what you have.

Comment (adam-winfree) [reply .gitignore:13]: Are you kidding? Everyone should have what I have 😄

File: src/Intacct/Functions/AccountsPayable/AbstractVendor.php [hunk 16]
```diff
@@ -232,6 +237,65 @@ abstract class AbstractVendor extends AbstractFunction
     /** @var array */
     protected $restrictedDepartments = [];
 
+    /** @var array */
+    protected $vendorDesignations = [];
+
+    /**
+     * @var
```
Comment (adam-winfree) [inline src/Intacct/Functions/AccountsPayable/AbstractVendor.php:244]: I'm just a var, take a good look at me. Just your typical prototype.

File: src/Intacct/Functions/AccountsPayable/AbstractVendor.php (hunk 8 above)
Comment (adam-winfree) [reply src/Intacct/Functions/AccountsPayable/AbstractVendor.php:1627]: And the bracket should be on the new line. Same for below

**Review by adam-winfree — CHANGES_REQUESTED** [review]
I have no idea how RealPage works, so my thoughts are pretty simple.

File: src/Intacct/ClientConfig.php [hunk 17]
```diff
@@ -254,6 +254,40 @@ public function setUserPassword(string $userPassword)
...
+     */
+    public function setHeaders(array $headers) : void
+    {
+        $this->headers = $headers;
+    }
+
+    /**
+     * @param string $key
+     * @param string $value
+     */
```
Comment (lancepioch) [inline src/Intacct/ClientConfig.php:278]:
```suggestion
    public function getHeaders() : ?array
    {
        return $this->headers;
    }

    public function setHeaders(array $headers) : void
    {
        $this->headers = $headers;
    }

```

Don't need docblocks!

File: src/Intacct/Functions/AccountsPayable/AbstractVendor.php [hunk 18]
```diff
@@ -232,6 +237,65 @@ abstract class AbstractVendor extends AbstractFunction
...
+    protected $vendorDesignations = [];
+
+    /**
+     * @var bool
+     */
+    protected $insuranceRequired;
+
+    /**
+     * @var
+     */
```
Comment (lancepioch) [inline src/Intacct/Functions/AccountsPayable/AbstractVendor.php:250]: I don't think these do anything but take up space?

**Review by dhazelett — APPROVED** [review]
